### ISSUE TITLE
Create a database library for registry integrity monitoring

### ISSUE THREAD

**Issue by antoniomanuelfr** [wazuh/wazuh#5867]
Hi team! 
The new implementation of the Windows registry monitoring needs a new structure for the FIM database. 

Two new tables will be created: 
- **registry_key**: This table will save the path of the key that is going to be monitored. This table has the following members: 
  - `path`: String with the path of the key.
  - `data_id:` Foreign key to the registry data table.
  - `perm`: String with the permissions of the key.
  - `uid`: Integer that specifies the user id of the key.
  - `gid`: Integer that specifies the group id of the key.
  - `user_name`: String that stores the user name of the key.
  - `group_name`: String that stores the group name of the key.
  - `scanned`: Integer value that indicates if the key was deleted in the previous scan.
  - `options`: Integer that stores the scan options that are going to be used in the key. 
- **registry_data**: This table will save all the values that are stored in the key. The alerts will be raised when any of the values inside a key is added, modified, or deleted. 
  - `key_id`: Foreign key to the registry key table.
  - `name`: String that stores the name of the value. 
  - `type`: Integer that stores the type of the value.
  - `scanned`: Integer value that indicates if the value was deleted in the previous scan.
  - `checksum`: Integer that stores the checksum.
  - `options`: Integer that stores the scan options that are going to be used in the value.
  - `last_event`: Time of the last event in the registry.

New functions must be developed in order to insert, modify, and delete keys from the DB.

The `entry_path` table and the `entry_data` table will be renamed to `file_entry` and `file_data` respectively and the `entry_type` key is going to be removed from the `file_entry` table.

New queries need to be created in order to: 
- Return the values of a key.
- Return the values of a not scanned key.
- Remove all the values of a key.
- Set all the values of a key to scanned.

**Comment by jesusjimsa** [comment]
The working branch is [5867-fim-registry-database](https://github.com/wazuh/wazuh/tree/5867-fim-registry-database).

**Comment by jesusjimsa** [comment]
The `entry_path` and `entry_data` tables have been renamed to `file_entry` and `file_data`. With them, I have also renamed two function headers (`fim_db_get_count_entry_data` and `fim_db_get_count_entry_path` to `fim_db_get_count_file_data` and `fim_db_get_count_file_entry`) and one structure (`fim_entry_data` to `fim_entry_data`).
Some wrappers used in the unit tests have been affected as well.

**Comment by jesusjimsa** [comment]
As we are going to remove `entry_type`, it will be removed from the functions that use it. We have decided that the ones that relay in its value to work properly will have a separate version for registries. This will affect for example `fim_configuration_directory` or `fim_json_event`.

**Comment by jesusjimsa** [comment]
Here are some queries we will need for the database:

- Get key attributes for an entry select by path
- Get data key attributes for an entry select by `key_id` and name
- Insert or replace `registry_key` entry
- Insert or replace `registry_data` entry
- Get `registry_key` not scanned
- Get `registry_data` not scanned
- Set `registry_key` as scanned
- Set `registry_data` as scanned
- Delete `registry_data` entry
- Delete `registry_key` entry and all `registry_data` referenced by `data_id`
- Get number of entries (keys + values)

**Comment by jesusjimsa** [comment]
Since _sqlite3_ does not allow deleting from multiple tables in one single statement, the function which performs this operation must use two different statements. First, it must delete all **registry_data** entries referenced by a `data_id` and then, delete the **registry_key** entry referenced by a path.

**Comment by antoniomanuelfr** [comment]
New functions will be added to de database. These are the headers of the functions that need to be added (this list can be changed)

- [ ] `int fim_db_get_registry_data_checksum(fdb_t *fim_sql, void * arg)`
- [ ] `int fim_db_get_registry_data_checksum(fdb_t *fim_sql, void * arg)`
- [ ] `fim_registry_data **fim_db_get_value_path(fdb_t *fim_sql, const char *key_path, const char *value)`
- [ ] `char **fim_db_get_values_from_registry_key(fdb_t *fim_sql, const char *key_path, const unsigned long int key_id)`
- [ ] `int fim_db_insert_registry_value(fdb_t *fim_sql, const char *key_path, fim_registry_data *data, int *row_id)`
- [ ] `int fim_db_insert_registry_key(fdb_t *fim_sql, const char *key_path, fim_registry_key *entry, int key_id)`
- [ ] `int fim_db_insert_registry(fdb_t *fim_sql, const char *key_path, fim_registry_data *new, fim_registry_data *saved)`
- [ ] `int fim_db_registry_key_checksum_range(fdb_t *fim_sql, const char *start, const char *top, const long id, const int n, pthread_mutex_t *mutex)`
- [ ] `int fim_db_get_registry_key_count_range(fdb_t *fim_sql, char *start, char *top, int *counter)`
- [ ] `int fim_db_get_registry_data_count_range(fdb_t *fim_sql, char *start, char *top, int *counter)`
- [ ] `void fim_db_remove_registry_key(fdb_t *fim_sql, fim_registry_key *key_entry, pthread_mutex_t *mutex, void *alert, void *fim_ev_mode, void *w_evt)`
- [ ] `void fim_db_remove_registry_value(fdb_t *fim_sql, char *key_path, fim_registry_data *key_entry, pthread_mutex_t *mutex, void *alert, void *fim_ev_mode, void *w_evt)`
- [ ] `int fim_db_get_row_registry_key(fdb_t *fim_sql, int mode, char **path)`
- [ ] `int fim_db_get_row_registry_data(fdb_t *fim_sql, int mode, char **path)`
- [ ] `int fim_db_set_all_registry_key_unscanned(fdb_t *fim_sql)`
- [ ] `int fim_db_set_all_registry_data_unscanned(fdb_t *fim_sql)`
- [ ] `int fim_db_set_registry_key_scanned(fdb_t *fim_sql, char *path)`
- [ ] `int fim_db_set_registry_value_scanned(fdb_t *fim_sql, char *name, unsigned int key_id)`
- [ ] `int fim_db_get_registry_keys_not_scanned(fdb_t * fim_sql, fim_tmp_file **file, int storage)`
- [ ] `int fim_db_get_registry_data_not_scanned(fdb_t * fim_sql, fim_tmp_file **file, int storage)`
- [ ] `int fim_db_delete_registry_keys_not_scanned(fdb_t *fim_sql, fim_tmp_file *file, pthread_mutex_t *mutex, int storage)`
- [ ] `int fim_db_get_registry_keys_range(fdb_t *fim_sql, char *start, char *top, fim_tmp_file **file, int storage)`
- [ ] `int fim_db_get_registry_value_range(fdb_t *fim_sql, char *start, char *top, fim_tmp_file **file, int storage)`
- [ ] `int fim_db_delete_registry_key_range(fdb_t * fim_sql, fim_tmp_file *file, pthread_mutex_t *mutex, int storage)`
- [ ] `int fim_db_delete_registry_value_range(fdb_t * fim_sql, fim_tmp_file *file, pthread_mutex_t *mutex, int storage)`
- [ ] `int fim_db_process_missing_registry_key_entry(fdb_t *fim_sql, fim_tmp_file *file, pthread_mutex_t *mutex, int storage, fim_event_mode mode, whodata_evt * w_evt)`
- [ ] `int fim_db_process_missing_registry_data_entry(fdb_t *fim_sql, fim_tmp_file *file, pthread_mutex_t *mutex, int storage, fim_event_mode mode, whodata_evt * w_evt)`
- [ ] `void fim_db_callback_save_registry_value(fdb_t *fim_sql, fim_registry_key *key, fim_registry_data *value, int storage, void *arg)`
- [ ] `int fim_db_get_count_registry_entry(fdb_t *fim_sql)`
- [ ] `int fim_db_get_count_registry_data(fdb_t * fim_sql)`
- [ ] `static registry_key *fim_db_decode_full_registry_row(sqlite3_stmt *stmt)`

The binding functions need to be implemented for each SQL statement.

- [ ] `static void fim_db_insert_reg_data(fdb_t *fim_sql, fim_registry_data *data)`
- [ ] `static fim_db_bind_replace_path(fdb_t *fim_sql, const char *key_path, int row_id, fim_registry_data *data`
- [ ] `static void fim_db_bind_registry_key(fdb_t *fim_sql, int index, const char *key_path)`
- [ ] `static void fim_db_bind_update_registry_data(fdb_t *fim_sql, fim_registry_data *entry, int *row_id)`
- [ ] `static void fim_db_bind_update_registry_key(fdb_t *fim_sql, const char *key_path, fim_registry_data *data`
- [ ] `static void fim_db_bind_set_registry_key_count(fdb_t *fim_sql, const char *path)`
- [ ] `static void fim_db_bind_get_registry_data_id(fdb_t *fim_sql, const char *path)`
- [ ] `static void fim_db_bind_get_registry_count_range(fdb_t *fim_sql, const char *start, const char *stop)`
- [ ] `static void fim_db_bind_get_registry_path_range(fdb_t *fim_sql, const char *start, const char *stop)`
- [ ] `static void fim_db_bind_delete_registry_key_path(fdb_t *fim_sql, const char *key_path`
- [ ] `static void fim_db_bind_delete_registry_data(fdb_t *fim_sql, unsigned int key_id, const char *value_name)`

**Comment by jesusjimsa** [comment]
We have added, as of now, 24 statements. We will divide the implementation of the 24 binding functions as follows:
Antonio
FIMDB_STMT_REPLACE_REG_DATA
FIMDB_STMT_REPLACE_REG_KEY
FIMDB_STMT_GET_REG_KEY
FIMDB_STMT_GET_REG_DATA
FIMDB_STMT_UPDATE_REG_DATA
FIMDB_STMT_UPDATE_REG_KEY
FIMDB_STMT_GET_ALL_REG_ENTRIES
FIMDB_STMT_GET_REG_KEY_NOT_SCANNED
FIMDB_STMT_GET_REG_DATA_NOT_SCANNED
FIMDB_STMT_SET_ALL_REG_KEY_UNSCANNED
FIMDB_STMT_SET_REG_KEY_UNSCANNED
FIMDB_STMT_SET_ALL_REG_DATA_UNSCANNED
@jesusjimsa 
- FIMDB_STMT_SET_REG_DATA_UNSCANNED
- FIMDB_STMT_GET_REG_DATA_ID
- FIMDB_STMT_DELETE_REG_KEY_PATH
- FIMDB_STMT_DELETE_REG_DATA
- FIMDB_STMT_DELETE_REG_DATA_PATH
- FIMDB_STMT_GET_COUNT_REG_KEY
- FIMDB_STMT_GET_COUNT_REG_DATA
- FIMDB_STMT_GET_COUNT_REG_KEY_AND_DATA
- FIMDB_STMT_GET_LAST_REG_KEY
- FIMDB_STMT_GET_FIRST_REG_KEY
- FIMDB_STMT_GET_REG_COUNT_RANGE
- FIMDB_STMT_GET_REG_PATH_RANGE

**Comment by jesusjimsa** [comment]
We have added, as of now, 24 statements. We will divide the implementation of the 24 binding functions as follows:

@antoniomanuelfr 

- [ ] `FIMDB_STMT_REPLACE_REG_DATA`
- [ ] `FIMDB_STMT_REPLACE_REG_KEY`
- [ ] `FIMDB_STMT_GET_REG_KEY`
- [ ] `FIMDB_STMT_GET_REG_DATA`
- [ ] `FIMDB_STMT_UPDATE_REG_DATA`
- [ ] `FIMDB_STMT_UPDATE_REG_KEY`
- [ ] `FIMDB_STMT_GET_ALL_REG_ENTRIES`
- [ ] `FIMDB_STMT_GET_REG_KEY_NOT_SCANNED`
- [ ] `FIMDB_STMT_GET_REG_DATA_NOT_SCANNED`
- [ ] `FIMDB_STMT_SET_ALL_REG_KEY_UNSCANNED`
- [ ] `FIMDB_STMT_SET_REG_KEY_UNSCANNED`
- [ ] `FIMDB_STMT_SET_ALL_REG_DATA_UNSCANNED`

@jesusjimsa 

- [ ] `FIMDB_STMT_SET_REG_DATA_UNSCANNED`
- [ ] `FIMDB_STMT_GET_REG_DATA_ID`
- [ ] `FIMDB_STMT_DELETE_REG_KEY_PATH`
- [ ] `FIMDB_STMT_DELETE_REG_DATA`
- [ ] `FIMDB_STMT_DELETE_REG_DATA_PATH`
- [ ] `FIMDB_STMT_GET_COUNT_REG_KEY`
- [ ] `FIMDB_STMT_GET_COUNT_REG_DATA`
- [ ] `FIMDB_STMT_GET_COUNT_REG_KEY_AND_DATA`
- [ ] `FIMDB_STMT_GET_LAST_REG_KEY`
- [ ] `FIMDB_STMT_GET_FIRST_REG_KEY`
- [ ] `FIMDB_STMT_GET_REG_COUNT_RANGE`
- [ ] `FIMDB_STMT_GET_REG_PATH_RANGE`

**Comment by jesusjimsa** [comment]
We missed two SQL statements: `FIMDB_STMT_SET_REG_KEY_SCANNED` and `FIMDB_STMT_SET_REG_DATA_SCANNED`. They have been added now to the `SQL_STMT` array and the bind functions.

**Comment by jesusjimsa** [comment]
After modifying the data structures, the bind functions for the SQL statements will be changed to use the `fim_entry` structure.

**Comment by jesusjimsa** [comment]
We have modified the SQL statements to use the new SQL schema added with #5873. We need to adapt the bind functions for these new statements.
After modifying the statements, we will change the file structure to separate the file and the registry functionality in `fim_db` (for common functions), `fim_db_files` (for functions that affect the files' database) and `fim_db_registry` (for functions that affect the registries' database).

**Comment by jesusjimsa** [comment]
I have modified the SQL statements and bindings to adapt them to the changes introduced in the schema and the structures in this commits:
* 56f5816e156ee5dd077bfb17a23d4554cf39ffd5
* 8ed0a1fc7d3b5fb90d249b5c37462b90f2c849d5
* 120fe28a5813163f3d621a6f54d264d713d5c533
* 1e47e0bfbdc3c26b56fe0907a0dfd0f89b379103

We have also created a PoC to test the functions we write. We spent almost an entire day fixing it, because it was an adaptation of an old PoC and had a lot of compilation errors. It is working now and should help us implement the functions without errors.
